src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+  const auto matching_assignment=std::find_if(
+      next_instr,
+      end,
+      [&return_identifier](const goto_programt::instructiont &instr)
+      {
+        return is_assignment_from(instr, return_identifier);
+      });
+
+  assert(matching_assignment!=end);
+
+  // Assume that the LHS of *this* assignment is the actual nondet variable
+  const auto &code_assign=to_code_assign(matching_assignment->code);
+  const auto nondet_var=code_assign.lhs();
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
+  // Generate nondet init code
+  code_blockt init_code;
+  gen_nondet_init(
+    nondet_var,
+    init_code,
+    symbol_table,
+    source_loc,
+    false,
+    true,
+    instr_info.is_never_null,
+    message_handler,
+    max_nondet_array_length);
+
+  // Convert this code into goto instructions
+  goto_programt new_instructions;
+  goto_convert(init_code, symbol_table, new_instructions, message_handler);
+
+  // Insert the new instructions into the instruction list
+  return insert_before(
+    instructions,
+    after_erased,
+    new_instructions.instructions);
+}
+
+/*******************************************************************\
+
+Function: remove_java_nondet
+
+  Inputs:
+    message_handler: Object which prints messages.
+    symbol_table: The list of known symbols.
+    max_nondet_array_length: The maximum length of new nondet arrays.
+    goto_program: The program to modify.
+
+ Purpose: Modify a goto_programt to replace 'nondet' library functions with
+          CBMC-native nondet expressions.
+
+\*******************************************************************/
+
+static void remove_java_nondet(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  const size_t max_nondet_array_length,
+  goto_programt &goto_program)
+{
+  // Check each instruction.
+  // `process_target` may modify the list in place, and returns the next
+  // iterator to look at.
+  for(auto instruction_iterator=std::begin(goto_program.instructions),
+        end=std::end(goto_program.instructions);
+      instruction_iterator!=end;
+      instruction_iterator=process_target(
+        message_handler,
+        symbol_table,
+        max_nondet_array_length,
+        goto_program.instructions,
+        instruction_iterator))
+  {
+    // Loop body deliberately empty.
+  }
+}
+
+void remove_java_nondet(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  const size_t max_nondet_array_length,
+  goto_functionst &goto_functions)
+{
+#ifdef JAVA_NONDET_DEBUG
+  std::cerr
+    << "replacing nondet statements using max array size of: "
+    << max_nondet_array_length
+    << '\n';
+#endif
+
+  for(auto &f : goto_functions.function_map)
+  {
```

Comment:
What's the advantage vs the range-for? forall_goto_functions hides whether the iteration is mutable or const, computes the end iterator on each iteration, and post-increments rather than pre-increments the iterator, which may lead to an additional copy per iteration. Range-for seems strictly better in terms of readability and performance.